File: app/hero.ts
import { Cell } from "./maze";

/**
 * Hero
 */
export class Hero {
    private _x: number = 0;
    private _y: number = 0;
    private maxLife = 100;
    private _life: number = this.maxLife;

    public get x(): number {
        return this._x;
    }

    public get y(): number {
        return this._y;
    }

    public get life() {
        return this._life;
    }

    public goTo(cell: Cell) {
        this._x = cell.col;
        this._y = cell.row;
    }

    public moveTo(cell: Cell) {
        this.goTo(cell);
        this._life--;
    }

    public isOn(cell: Cell) {
        return this._x == cell.col && this._y == cell.row;
    }

    public changeLife(content: number) {
        this._life += content;
        if (this._life > this.maxLife) this._life = this.maxLife;
        if (this._life < 0) this._life = 0;
    }
}

File: src/hero.rs

/**
 * Hero
 */
struct Hero {
    pos: Position,
    maxLife: u32,
    life: u32
}

impl Hero {
    // pub fn x(): number {
    //     return self._x;
    // }

    // public get y(): number {
    //     return self._y;
    // }

    pub fn life() -> u32 {
        self.life;
    }

    pub fn go_to(room: &Room) {
        self._x = cell.col;
        self._y = cell.row;
    }

    pub fn move_to(cell: Cell) {
        self.go_to(cell);
        self.life -= 1;
    }

    pub fn isOn(cell: Cell) {
        return self._x == cell.col && self._y == cell.row;
    }

    pub fn change_life(&mut self, content: number) {
        self._life += content;
        if (self._life > self.maxLife) {self._life = self.maxLife};
        if (self._life < 0) {self._life = 0;}
    }
}
